File: graphs (undirected + directed)/ArticulationPoint.java
package graphs.revision;

public class ArticulationPoint extends GraphTraversal {

	int reachableAncestor[];
	int treeOutDegree[];

	ArticulationPoint(int vertices) {
		super(vertices);
		reachableAncestor = new int[vertices + 1];
		treeOutDegree = new int[vertices + 1];
	}

	public static void main(String[] args) {
		int vertices = 6;
		Graph graph = new Graph(vertices, false);

		graph.insertEdge(1, 2);
		graph.insertEdge(1, 5);
		graph.insertEdge(1, 6);

		graph.insertEdge(2, 5);
		graph.insertEdge(2, 3);

		graph.insertEdge(3, 4);
		graph.insertEdge(4, 5);

		ArticulationPoint dfs = new ArticulationPoint(vertices);

		dfs.depthFirstSearchRecursive(graph, 1);

	}

	public void processVertexEarly(int vertex) {
		reachableAncestor[vertex] = vertex;
	}

	public void processEdge(int x, int y) {
		String type = edgeClassification(x, y);
		if (type.equals("TREE"))
			treeOutDegree[x] = treeOutDegree[x] + 1;
		if (type.equals("BACK"))
			reachableAncestor[x] = y;
	}

	public void processVertexLate(int vertex) {
		exitTime[vertex] = ++time;

		boolean root;
		int timeV;
		int timeParent;

		if (parents[vertex] < 1) {
			if (treeOutDegree[vertex] > 1)
				System.out.printf("root articulation vertex: %d \n", vertex);
			return;
		}
		root = (parents[parents[vertex]] < 1); /* test if parent[v] is root */
		if (!root) {
			if (reachableAncestor[vertex] == parents[vertex]) {
				System.out.printf("parent articulation vertex: %d \n", parents[vertex]);
			}
			if (reachableAncestor[vertex] == vertex) {
				System.out.printf("bridge articulation vertex: %d \n", parents[vertex]);
				if (treeOutDegree[vertex] > 0) /* test if v is not a leaf */
					System.out.printf("bridge articulation vertex: %d \n", vertex);
			}
		}
		timeV = entryTime[reachableAncestor[vertex]];
		timeParent = entryTime[reachableAncestor[parents[vertex]]];
		if (timeV < timeParent)
			reachableAncestor[parents[vertex]] = reachableAncestor[vertex];
	}

	public String edgeClassification(int x, int y) {
		if (parents[y] == x)
			return "TREE";
		if (discovered[y] && !processed[y])
			return "BACK";
		if (processed[y] && (entryTime[y] > entryTime[x]))
			return "FORWARD";
		if (processed[y] && (entryTime[y] < entryTime[x]))
			return "CROSS";

		System.out.printf("Warning: unclassified edge (%d,%d)\n", x, y);
		return "";
	}

}

File: implementations/graphs/articulationPoint.mjs
import { Graph } from "../../data-structures/Graph.mjs";
import DFS from "../../algorithms/graphs/DFS.mjs";

const vertices = 6;
const g = new Graph(vertices, false);
g.insertEdges([[1, 6],[1, 5],[1, 2],[2, 5],[2, 3],[3, 4],[4, 5]]);

const dfs = new DFS(g);

dfs.processVertexEarly = function (v) {
  reachableAncestor[v] = v;
};

dfs.processEdge = function (x, y) {
  if (this.parents[y] == x) {
    treeOutDegree[x] = treeOutDegree[x] + 1;
  }

  if (this.discovered[y] && !this.processed[y] && this.parents[x] !== y) {
    if (this.entryTime[y] < this.entryTime[reachableAncestor[x]]) {
      reachableAncestor[x] = y;
    }
  }
};

dfs.processVertextLate = function (v) {
  let isRoot;
  let timeV;
  let timeParent;

  if (this.parents[v] < 1) {
    if (treeOutDegree[v] > 1) {
      console.log(`root articulation vertex: ${v}`);
    }
    return;
  }

  isRoot = (this.parents[this.parents[v]] < 1);

  if (!isRoot) {
    if (reachableAncestor[v] == this.parents[v]) {
      console.log(`parent articulation vertex: ${this.parents[v]}`);
    }
    if (reachableAncestor[v] == v) {
      console.log(`bridge articulation vertex: ${this.parents[v]}`);
      if (treeOutDegree[v] > 0)
        console.log(`bridge articulation vertex: ${v}`);
    }
  }
  timeV = this.entryTime[reachableAncestor[v]];
  timeParent = this.entryTime[reachableAncestor[this.parents[v]]];
  if (timeV < timeParent) reachableAncestor[this.parents[v]] = reachableAncestor[v];
};

const reachableAncestor = new Array(vertices + 1);
const treeOutDegree = new Array(vertices + 1).fill(0);
dfs.parents[1] = -1;
dfs.search(1);
